### ISSUE TITLE
Basic type narrowing / flow analysis

### ISSUE THREAD

**Issue by LoganDark** [Benjamin-Dobell/IntelliJ-Luanalysis#127]
Currently, no type narrowing / flow analysis is done in Luanalysis (as far as I can tell), which means I constantly have to annotate expressions with `--[[--- @not nil ]]` even if I *just did a nil check*. Example:

```lua
--- @type number | nil
local x = 5

if x then
	math.max(x, 5) -- warning! nil is not assignable to number
end
```

Luanalysis should be able to detect these cases and deduce that `x` is neither `false` nor `nil`. That means `nil` will be removed from ~~type intersections~~ ~~type unions~~ *the `|` thing* and `boolean` will be converted to an explicit `true`.

Note that this does not extend to table members or global variables, because those may have metatables. Therefore this should always be a warning:

```lua
--- @shape Something
--- @field x number | nil

--- @type Something
local x = {x = 5}

if x.x then
	math.max(x.x, 5) -- warning! x.x could have changed in the meantime
end
```

But you can still store `x.x` in a local so that the local itself does not change:

```lua
--- @shape Something
--- @field x number | nil

--- @type Something
local x = {x = 5}
local xx = x.x

if xx then
	math.max(xx, 5) -- ok! x.x could have changed in the meantime, but xx is fine
end
```

This is more complex than it seems though, because functions with upvalues can still change them. So upvalues need to be automatically detected & calls to those functions identified. But *that's* also more complicated than it seems, because functions are first-class citizens, so you don't necessarily know if a function you call is either one that has a certain upvalue, or one that transitively calls one of those. Here are some theoretical scenarios, with explanations...

```lua
--- @type number | nil
local x = 5

local function changeX()
	x = nil
end

if x then
	-- `changeX` was defined, but it wasn't hooked into anything, and it's local, so it can't possibly have been called by now.
	-- `x` is type-narrowed.
	math.max(x, 5) -- ok!
end
```

```lua
--- @type number | nil
local x = 5

function changeX()
	x = nil
end

if x then
	-- `changeX` was defined globally. The global env may have a __newindex metamethod - all bets are off.
	math.max(x, 5) -- warning: x may have changed in the meantime; changeX function is accessible
end
```

```lua
--- @type number | nil
local x = 5

local function changeX()
	x = nil
end

if x then
	-- `changeX` is defined locally and not hooked into anything, but we call it:
	changeX()
	math.max(x, 5) -- warning: x may have changed in the meantime; changeX function mutates it as an upvalue
	-- More advanced flow analysis (like detecting functions which conditionally modify upvalues) is... more difficult

	if x then
		math.max(x, 5) -- ok! changeX was not called
	end
end
```

```lua
--- @type number | nil
local x = 5

local function changeX()
	x = nil
end

registerSomething(changeX)

if x then
	-- `changeX` is defined locally, but got hooked into something
	math.max(x, 5) -- warning: x may have changed in the meantime; changeX function may have been called
	-- for all we know, `math.max` uses the `changeX` closure that we passed to `registerSomething`
	-- once again, all bets are off
end
```

Of course, there are more issues, like detecting whether a particular scope executes after `changeX` has been shared with something else. Sometimes it's both. You really have to be pessimistic here, and can't make any assumptions, which is really sad for ease of implementation, but it'll be better in the long run.

This is a large and difficult feature to add. I do not have any implementation to contribute and this issue will probably stay open for a good long while (or forever). But I feel it's worth tracking. Especially with the test cases.

**Comment by Benjamin-Dobell** [comment]
There are already issues https://github.com/Benjamin-Dobell/IntelliJ-Luanalysis/issues/4, https://github.com/Benjamin-Dobell/IntelliJ-Luanalysis/issues/63 and closed issues https://github.com/Benjamin-Dobell/IntelliJ-Luanalysis/issues/100 and https://github.com/Benjamin-Dobell/IntelliJ-Luanalysis/issues/97.

I don't want too many duplicate issues so am closing this issue. Although this isn't new to me here (I touched on some of this in #4), I'd suggest copying this content into a comment on #63, so it's not lost. That should also make it easier to find #63 with search, since it'll then include both smart casting and narrowing terminology.

**Comment by LoganDark** [comment]
> There are already issues #4, #63 and closed issues #100 and #97.

jesus, I must be the biggest idiot ever. :/

I thought I remembered looking at something like that but I couldn't remember the name and whether or not it was on the EmmyLua repository instead. Thanks.

> I'd suggest copying this content into a comment on #63

The references you made by linking the issues should be enough (this one will appear on all of those).
